pkg/cortexpb/compat.go

```diff
@@ -0,0 +1,235 @@
+// It uses unsafe, but as LabelAdapter == labels.Label this should be safe.
+// This allows us to use labels.Labels directly in protos.
+func FromLabelsToLabelAdapters(ls labels.Labels) []LabelAdapter {
+	return *(*[]LabelAdapter)(unsafe.Pointer(&ls))
+}
+
+// FromLabelAdaptersToMetric converts []LabelAdapter to a model.Metric.
+// Don't do this on any performance sensitive paths.
+func FromLabelAdaptersToMetric(ls []LabelAdapter) model.Metric {
+	return util.LabelsToMetric(FromLabelAdaptersToLabels(ls))
+}
+
+// FromMetricsToLabelAdapters converts model.Metric to []LabelAdapter.
+// Don't do this on any performance sensitive paths.
+// The result is sorted.
+func FromMetricsToLabelAdapters(metric model.Metric) []LabelAdapter {
+	result := make([]LabelAdapter, 0, len(metric))
+	for k, v := range metric {
+		result = append(result, LabelAdapter{
+			Name:  string(k),
+			Value: string(v),
+		})
+	}
+	sort.Sort(byLabel(result)) // The labels should be sorted upon initialisation.
+	return result
+}
+
+type byLabel []LabelAdapter
+
+func (s byLabel) Len() int           { return len(s) }
+func (s byLabel) Less(i, j int) bool { return strings.Compare(s[i].Name, s[j].Name) < 0 }
+func (s byLabel) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+
+var json = jsoniter.ConfigCompatibleWithStandardLibrary
+
+// MarshalJSON implements json.Marshaler.
+func (s Sample) MarshalJSON() ([]byte, error) {
+	t, err := json.Marshal(model.Time(s.TimestampMs))
+	if err != nil {
+		return nil, err
+	}
+	v, err := json.Marshal(model.SampleValue(s.Value))
+	if err != nil {
+		return nil, err
+	}
+	return []byte(fmt.Sprintf("[%s,%s]", t, v)), nil
+}
+
+// UnmarshalJSON implements json.Unmarshaler.
+func (s *Sample) UnmarshalJSON(b []byte) error {
+	var t model.Time
+	var v model.SampleValue
+	vs := [...]stdjson.Unmarshaler{&t, &v}
+	if err := json.Unmarshal(b, &vs); err != nil {
+		return err
+	}
+	s.TimestampMs = int64(t)
+	s.Value = float64(v)
+	return nil
+}
+
+// MetricMetadataMetricTypeToMetricType converts a metric type from our internal client
+// to a Prometheus one.
+func MetricMetadataMetricTypeToMetricType(mt MetricMetadata_MetricType) textparse.MetricType {
+	switch mt {
+	case UNKNOWN:
+		return textparse.MetricTypeUnknown
+	case COUNTER:
+		return textparse.MetricTypeCounter
+	case GAUGE:
+		return textparse.MetricTypeGauge
+	case HISTOGRAM:
+		return textparse.MetricTypeHistogram
+	case GAUGEHISTOGRAM:
+		return textparse.MetricTypeGaugeHistogram
+	case SUMMARY:
+		return textparse.MetricTypeSummary
+	case INFO:
+		return textparse.MetricTypeInfo
+	case STATESET:
+		return textparse.MetricTypeStateset
+	default:
+		return textparse.MetricTypeUnknown
+	}
+}
+
+// Only set from tests to get special behaviour to verify that custom marshal and unmarshal is used.
+var isTesting = false
+
+func init() {
+
+	jsoniter.RegisterTypeEncoderFunc("cortexpb.Sample", func(ptr unsafe.Pointer, stream *jsoniter.Stream) {
```

Comment:
We've changed `client.Sample` into `cortexpb.Sample`. Is it backward compatible?

Reply:
This string is used to match type name with correct encoder/decoder when using jsoniter library. I've added tests to "old" `compat` package to verify that custom marshal/unmarshal functions are used when working with `client.Sample` type alias.